File: Python/afe/algorithms/sorting/shell_sort.py
# -*- coding: utf-8 -*-

from typing import List, TypeVar, Optional, Callable
from .sort_interface import SortInterface


T = TypeVar('T')


class ShellSort(SortInterface):

    """
    A class that implements the Shell Sort algorithm. \n
    Shell Sort is an in-place comparison sort that generalizes insertion sort to allow the exchange of items that are far apart.
    It is not a stable sort, meaning that it does not maintain the relative order of records with equal keys.
    """

    @staticmethod
    def sort(data: List[T], comparator: Optional[Callable[[T, T], int]] = None) -> None:
        
        """
        Sorts a list in-place using Shell Sort. Not stable.
        Uses Knuth's gap sequence.
        - Time Complexity: O(n log^2 n) to O(n^1.5)
        - Space Complexity: O(1)
        """

        if comparator is None:
            comparator = lambda a, b: (a > b) - (a < b)

        n = len(data)
        gap = 1
        while gap < n / 3:
            gap = 3 * gap + 1

        while gap >= 1:
            for i in range(gap, n):
                temp = data[i]
                j = i
                while j >= gap and comparator(data[j - gap], temp) > 0:
                    data[j] = data[j - gap]
                    j -= gap
                data[j] = temp
            gap = gap // 3

File: Java/src/main/java/io/github/mihaistreames/afe/algorithms/sorting/ShellSort.java
package io.github.mihaistreames.afe.algorithms.sorting;

import org.jetbrains.annotations.NotNull;

import java.util.Comparator;
import java.util.List;

/**
 * An in-place comparison sort. It can be seen as either a generalization of
 * sorting by exchange (bubble sort) or sorting by insertion (insertion sort).
 * <p>
 * This implementation is not stable. <br>
 * Time Complexity: O(n log^2 n) is common, depends on gap sequence. <br>
 * Space Complexity: O(1)
 */
public final class ShellSort {

    private ShellSort() {
        throw new UnsupportedOperationException("Utility class cannot be instantiated");
    }

    /**
     * Sorts a list using the shell sort algorithm and natural ordering.
     *
     * @param <T>  the type of elements, must extend {@link Comparable}
     * @param list the list to be sorted
     */
    public static <T extends Comparable<T>> void shellSort(@NotNull final List<T> list) {
        sort(list, Comparator.naturalOrder());
    }

    /**
     * Sorts a list using the shell sort algorithm and a custom comparator.
     *
     * @param <T>        the type of elements
     * @param list       the list to be sorted
     * @param comparator the comparator to determine the order
     */
    public static <T> void shellSort(@NotNull final List<T> list,
                                     @NotNull final Comparator<T> comparator) {
        sort(list, comparator);
    }

    private static <T> void sort(@NotNull final List<T> list,
                                 @NotNull final Comparator<T> comparator) {
        final int n = list.size();
        for (int gap = n / 2; gap > 0; gap /= 2) {
            for (int i = gap; i < n; i++) {
                final T temp = list.get(i);
                int j;
                for (j = i; j >= gap && comparator.compare(list.get(j - gap), temp) > 0; j -= gap) {
                    list.set(j, list.get(j - gap));
                }
                list.set(j, temp);
            }
        }
    }
}